src/Microsoft.AspNet.Mvc.ViewFeatures/Buffer/ViewBufferTextWriter.cs

```diff
@@ -0,0 +1,172 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNet.Mvc.ViewFeatures.Buffer
+{
+    public class ViewBufferTextWriter : TextWriter
+    {
+        private const int PageSize = 1024;
+
+        private readonly TextWriter _inner;
+        private readonly List<char[]> _pages;
+        private readonly ArrayPool<char> _pool;
+
+        private int _currentPage;
+        private int _currentIndex; // The next 'free' character
+
+        public ViewBufferTextWriter(ArrayPool<char> pool, TextWriter inner)
+        {
+            _pool = pool;
+            _inner = inner;
+            _pages = new List<char[]>();
+        }
+
+        public override Encoding Encoding => _inner.Encoding;
+
+        public override void Flush()
+        {
+            // Don't do anything. We'll call FlushAsync.
+        }
+
+        public override async Task FlushAsync()
+        {
+            if (_pages.Count == 0 || (_currentPage == 0 && _currentIndex == 0))
```

Comment:
You could generalize the second condition into the loop if it's a common case `if (i == _currentPage && _currentIndex == 0)`